src/prepared_statement.c

```diff
@@ -0,0 +1,339 @@
+#include "bouncer.h"
+
+#include <usual/crypto/csrandom.h>
+#include <usual/hashtab-impl.h>
+
+/*
+ * Like uthash HASH_DELETE, but doesn't remove item from hash table just
+ * unlink element from the doubly-linked-list.
+ */
+#define HASH_UNLINK(hh, head, delptr)									\
+	do {																\
+		struct UT_hash_handle *_hd_hh_del = &(delptr)->hh;				\
+		if (_hd_hh_del->prev != NULL) {									\
+			HH_FROM_ELMT((head)->hh.tbl, _hd_hh_del->prev)->next = _hd_hh_del->next; \
+		} else {														\
+			DECLTYPE_ASSIGN(head, _hd_hh_del->next);					\
+		}																\
+		if (_hd_hh_del->next != NULL) {									\
+			HH_FROM_ELMT((head)->hh.tbl, _hd_hh_del->next)->prev = _hd_hh_del->prev; \
+		} else {														\
+			_hd_hh_del->tbl->tail = HH_FROM_ELMT(_hd_hh_del->tbl, _hd_hh_del->prev); \
+		}																\
+	} while (0)
+
+static PgParsedPreparedStatement *create_prepared_statement(PgParsePacket *parse_packet)
+{
+	PgParsedPreparedStatement *s = (PgParsedPreparedStatement *)malloc(sizeof(*s));
+	s->name = parse_packet->name;
+	s->pkt = parse_packet;
+	return s;
+}
+
+static PgServerPreparedStatement *create_server_prepared_statement(PgSocket *server, PgParsedPreparedStatement *ps)
+{
+	PgServerPreparedStatement *s = NULL;
+	size_t query_len = strlen(ps->pkt->query);
+	size_t num_parameters = ps->pkt->num_parameters;
+	s = (PgServerPreparedStatement *)malloc(sizeof(*s) + query_len + 1 + num_parameters*4);
+	s->name_len = snprintf(s->name, sizeof(s->name), "B_%ld", server->nextUniquePreparedStatementID);
```

Comment:
According to the profile in another comment thread, this `snprintf` can take a serious amount of time.

Reply:
Can you sen me your pgbouncer.cfg and tell pgbench results you get?
I still can not reproduce the behaviour you observed: neither at my notebook, neither at EC2 host.

Reply:
I compile pgbouncer using:
```
./configure --with-cares --enable-cassert --with-systemd
make -j20
```

pgbouncer.ini
```ini
[databases]
postgres = host=localhost dbname=postgres

[pgbouncer]
pool_mode=transaction
listen_addr=127.0.0.1
auth_type=trust
admin_users=jelte
auth_file=auth_file.conf
query_wait_timeout=5
disable_prepared_statement_support=0
default_pool_size=30
log_connections=0
log_disconnections=0
```
auth_file.conf
```
"jelte" "password" ...
```
pgbouncer command
```bash
./pgbouncer pgbouncer.ini
```

custom-script.sql
```sql
\set aid random(1, 100000 * :scale)
SELECT aid, filler, bid, abalance, aid, filler, bid, abalance, aid, filler, bid, abalance, aid, filler, bid, abalance FROM pgbench_accounts WHERE aid = :aid;
```

pgbench command:
```bash
pgbench -T 10  -c 20 -j 20 -p6432 --protocol prepared -f custom-script.sql
```

Right after pgbench starts, in another run in another shell (this runs for 5 seconds and pgbench for 10, so make sure you start it quickly so `perf record` finishes before pgbench stops sending traffic):
```
perf record -F 300 -p {process_id_of_pgbouncer} --call-graph dwarf sleep 5
```

Then finally I look at the perf results using:
```
perf report
```

The pgbench results are as follows:
```
pgbench (15.2, server 12.14 (Ubuntu 12.14-1.pgdg22.04+1))
starting vacuum...end.
transaction type: custom-script.sql
scaling factor: 1
query mode: prepared
number of clients: 20
number of threads: 20
maximum number of tries: 1
duration: 10 s
number of transactions actually processed: 202835
number of failed transactions: 0 (0.000%)
latency average = 0.986 ms
initial connection time = 2.882 ms
tps = 20287.565628 (without initial connection time)
```

Because the runs are quite short and perf is a sampling profiler there's some serious deviation in the exact percentage. I tried a few runs and I get anywhere from ~3% to ~7% time spent in the direct body of `get_prepared_statement`. 

The perf report you shared looks as if perf is not using your debug symbols. If you would install binaries they are sometimes not there, and you need to install them separately. But when compiling from source they should be present by default. It is very important to use `--call-graph dwarf` on `perf record` to get useful results.

Reply:
I think the difference in TPS doesn't matter so much. It could be related to many things on my or your dev setup. One possibility is that I'm running on WSL, so effectively a Linux VM running on Windows, which could slow some stuff down. Also to be clear the core that's running pgbouncer is getting maxed out (even though most of my other cores are pretty idle).

What I do think is important is that we're able to get somewhat comparable `perf` output. Did you get more reasonable output this time, where time is actually spent in pgbouncer code? If not, can you share your output from `./configure`? Also what OS are you using?

Reply:
> What I do think is important is that we're able to get somewhat comparable perf output. Did you get more reasonable output this time, where time is actually spent in pgbouncer code? If not, can you share your output from ./configure? Also what OS are you using?

Yes, concrete performance numbers are not important.
But so big gap in performance can explain why we see completely different picture in perf profile.
If I observe such behavior just at y note, then most likely it means some problems with my  environment (although the fact that performance at my host is 3-x times higher makes me think that there are some problems at your side:) But I see exactly the same behavior at EC2 host.
For some reasons cost of system calls and context switches are much smaller  at your system comparing with speed of execution code in user space. It is really strange. in theory virtualization should omre affect time of system calls rather than speed of execution of normal instructions.

My note configuration:
```
Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz
RAM: 16GB
Linux xps 5.15.0-70-generic #77-Ubuntu SMP Tue Mar 21 14:02:37 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux
```
EC2 host: c5ad.2xlarge
```
AMD EPYC 7R32
RAM: 16GB
Linux ip-172-31-30-176 5.10.0-14-cloud-amd64 #1 SMP Debian 5.10.113-1 (2022-04-29) x86_64 GNU/Linux
```
Output of configure:
```
./configure 
checking build system type... x86_64-pc-linux-gnu
checking host system type... x86_64-pc-linux-gnu
checking target host type... unix
checking for gcc... gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether the compiler supports GNU C... yes
checking whether gcc accepts -g... yes
checking for gcc option to enable C11 features... none needed
checking for a BSD-compatible install... /usr/bin/install -c
checking for grep that handles long lines and -e... /usr/bin/grep
checking for a race-free mkdir -p... /usr/bin/mkdir -p
checking how to run the C preprocessor... gcc -E
checking whether compiler supports __func__... yes
checking whether linker supports --as-needed... yes
checking whether compiler supports dependency generation... yes
checking for working warning switches... done
checking whether ln -s works... yes
checking for egrep... /usr/bin/grep -E
checking for gawk... gawk
checking for a sed that does not truncate output... /usr/bin/sed
checking for strip... strip
checking for ranlib... ranlib
checking for ar... ar
checking for pkg-config... /usr/bin/pkg-config
checking pkg-config is at least version 0.9.0... yes
checking for pandoc... pandoc
checking for python3... python3
checking for strip... (cached) strip
checking for stdio.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for strings.h... yes
checking for sys/stat.h... yes
checking for sys/types.h... yes
checking for unistd.h... yes
checking for inttypes.h... (cached) yes
checking for stdbool.h... yes
checking for unistd.h... (cached) yes
checking for sys/time.h... yes
checking for sys/socket.h... yes
checking for poll.h... yes
checking for sys/un.h... yes
checking for arpa/inet.h... yes
checking for netinet/in.h... yes
checking for netinet/tcp.h... yes
checking for sys/param.h... yes
checking for sys/uio.h... yes
checking for pwd.h... yes
checking for grp.h... yes
checking for sys/wait.h... yes
checking for sys/mman.h... yes
checking for syslog.h... yes
checking for netdb.h... yes
checking for dlfcn.h... yes
checking for err.h... yes
checking for pthread.h... yes
checking for endian.h... yes
checking for sys/endian.h... no
checking for byteswap.h... yes
checking for malloc.h... yes
checking for regex.h... yes
checking for getopt.h... yes
checking for fnmatch.h... yes
checking for langinfo.h... yes
checking for xlocale.h... no
checking for linux/random.h... yes
checking for ucred.h... no
checking for sys/ucred.h... no
checking for sys/resource.h... yes
checking for sys/wait.h... (cached) yes
checking for inline... inline
checking for C/C++ restrict keyword... __restrict__
checking whether byte ordering is bigendian... no
checking for special C compiler options needed for large files... no
checking for _FILE_OFFSET_BITS value needed for large files... no
checking for pid_t... yes
checking for uid_t in sys/types.h... yes
checking for size_t... yes
checking for uint8_t... yes
checking for uint32_t... yes
checking for uint64_t... yes
checking for gcc options needed to detect all undeclared functions... none needed
checking whether strerror_r is declared... yes
checking for strlcpy... no
checking for strlcat... no
checking for strnlen... yes
checking for strsep... yes
checking for getpeereid... no
checking for sigaction... yes
checking for sigqueue... yes
checking for memmem... yes
checking for memrchr... yes
checking for mempcpy... yes
checking for inet_ntop... yes
checking for inet_pton... yes
checking for poll... yes
checking for getline... yes
checking for regcomp... yes
checking for err... yes
checking for errx... yes
checking for warn... yes
checking for warnx... yes
checking for getprogname... no
checking for setprogname... no
checking for posix_memalign... yes
checking for memalign... yes
checking for valloc... yes
checking for explicit_bzero... yes
checking for memset_s... no
checking for reallocarray... yes
checking for getopt... yes
checking for getopt_long... yes
checking for getopt_long_only... yes
checking for fls... no
checking for flsl... no
checking for flsll... no
checking for ffs... yes
checking for ffsl... yes
checking for ffsll... yes
checking for fnmatch... yes
checking for mbsnrtowcs... yes
checking for nl_langinfo... yes
checking for strtod_l... yes
checking for strtonum... no
checking for asprintf... yes
checking for vasprintf... yes
checking for timegm... yes
checking for localtime_r... yes
checking for gettimeofday... yes
checking for recvmsg... yes
checking for sendmsg... yes
checking for usleep... yes
checking for getrusage... yes
checking for syslog... yes
checking for mmap... yes
checking for getpeerucred... no
checking for arc4random_buf... no
checking for getentropy... yes
checking for getrandom... yes
checking for library containing WSAGetLastError... no
checking whether strerror_r returns char *... yes
checking for integer enc/dec functions... not found
checking for library containing clock_gettime... none required
checking for library containing getsockname... none required
checking for library containing gethostbyname... none required
checking for library containing hstrerror... none required
checking for lstat... yes
checking for libevent... yes
checking whether to build with systemd support... no
checking whether to use c-ares for DNS lookups... auto
checking for libcares >= 1.6.0... no
checking whether to use libudns... no
checking whether to use libevent for DNS lookups... yes
checking for OpenSSL... found
checking for SSL_CTX_use_certificate_chain_mem... no
checking for SSL_CTX_load_verify_mem... no
checking for asn1_time_parse... no
checking for root CA certs... /etc/ssl/certs/ca-certificates.crt
checking whether to build debug binary... yes
checking whether to enable asserts... no
checking whether to fail on warnings... no
configure: creating ./config.status
config.status: creating config.mak
config.status: creating lib/usual/config.h
config.status: lib/usual/config.h is unchanged

Results:
  adns    = evdns2
  pam     = no
  systemd = no
  tls     = yes
```

Reply:
> According to the profile in another comment thread, this `snprintf` can take a serious amount of time.

I see this happen as well on my Mac M1. All though I don't consider this chip as a benchmark for a real pgbouncer production setup.

@JelteF are we allowed, in regard to the license/copyright, to borrow this [conversion/formatting function](https://github.com/postgres/postgres/blob/master/src/backend/utils/adt/numutils.c#L891) from the postgresql server code?

Reply:
> are we allowed, in regard to the license/copyright, to borrow this [conversion/formatting function](https://github.com/postgres/postgres/blob/master/src/backend/utils/adt/numutils.c?rgh-link-date=2023-05-31T09%3A10%3A46Z#L891) from the postgresql server code?

Yes we're allowed, at least with a mention of the original copyright. PostgreSQL license is very liberal like BSD or MIT. But before doing that I think we should reorder the function calls, to simply call the function less often as I suggested in my second comment in this thread:

> Actually I guess a big part of the problem here is that create_server_prepared_statement is called even if the entry is found in server_prepared_statements in the get_prepared_statement. Instead create_prepared_statement should only be called when link_ps is NULL.

Reply:
@knizhnik I reached out to you over email to try and set up a screen share. I think that will be most efficient to debug this difference in profiling results.

Reply:
I still not able to reproduce your profile.
And frankly speaking it is a but suspicious to me that hash calculation or snprintf can take more CPU time that sending data through the sockets (even if it is local Unix sockets).

In any case - I committed new version where I have implemented you suggestion - with global hash table for query text. So now:
1. Query text is stored only once
2. snprint is done only once per query
3. Access to registered prepared statement doesn't require calculation of query hash.

Reply:
Looks good. Some initial perf results also look good to me, but it turned out my perf setup had some issues. So I'm planning to do some more perf testing next week.